### ISSUE TITLE
Failing mass balance & effects of implicit equation terms

### ISSUE THREAD

**Issue by achennu** [usnistgov/fipy#558]
I have been working with fipy to solve the equations for a photochemical reaction system. Firstly, I have to say that I love the library! It's a great piece of software. Thank you for sharing.

Having set up equations for a few reactants, I ran sweep() on the coupled equations and the temporal evolution of the reactant distribution, the source terms, etc looks logical. However, when I checked the mass balance of the equations I noticed that it was not correct. I mean that the quantity of the variable is many many orders of magnitude off from what I'd expect, despite the temporal evolution and shapes of the distribution profiles looking correct. I suspect this has something to do with Implicit source and diffusion term, or a misformulation on my part. I have been digging in to this to identify the configuration where the issue occurs. As a result I've created a script (long, but included here) that can be used to explore it interactively. If we resolve it, it could be adapted to an example in fipy eventually.

So the basic model setup is:
- Model is a 1D case of diffusion-reaction system (so pretty simple for fipy's scope)
- The domain has about 100 cells, with an 'active' mass located somewhere in it. For the simple case here, it is static i.e. it does not move or grow or reduce throughout the simulation.
- The light field is represented simply as a fraction of the intensity at the top of the domain. The domain has a base attenuation coefficient and the active mass has a different one. By combining them, the light field is set as `light = light0 * attenuation` to get exponential decay from the surface value `light0`.
- There is a main reactant `X` and another reactant `Y`, each of which have an equation. For the purposes of this example, `Y` equation is only a transient-diffusion equation with no sources terms. The reactant `X` is produced at a certain rate `Q` as long as the light value is in an optimum range.
- The simulation runs for a duration of `TOTAL`, the first half of which the light0 is 100 and then 0 for the next half.

Model source term variations in increasing complexity:
- The simplest case "production" means that when the light is on, the source term is `mass x Q x optimum(light)`. This is the basic production model, which we can call `prod0`.
- The next case is "production_consume", where an additional term occurs, where `X` is consumed at a constant rate `Q2` in the domain. So source terms: `prod0 - Q2`.
- The case `"production:inhibit[X]"` includes an auto-inhibition function. That is the production of `X` is modulated by a unitless function of its own concentration. So source term: `prod0 x inhibit(X)`
- The example includes more complex cases where the inhibition depends on `Y` or on both `X` and `Y`, etc but it is not necessary for the basic discussion here.

To estimate the quantitative accuracy of the equation solution, I make the following calculations for each small time step of the simulation:
- change from sources = depth_integration(sum(source_expressions))*dt giving units of mol/m**2. Can be thought of the net density of production/consumption within the domain.
- change from transport = diffusive transport loss at the boundary edges. This is approximated from fick's law: -D * dC/dx at the two edge cells, multiplied by the timestep. Units: mol/m**2
- Expected change at this timestep = source_changes - transport_changes
- So by calculating and accumulating this change over each timestep, I get an estimation for what the density of the variable X in the domain should be, in mol/m**2
- The actual density of X in the domain can be got at any timepoint by depth_integration(X), which gives mol/m**2 too
- The budget difference is then (expected - actual)

What I mean by the mass balance not seeming to be correct:
- When the light is on, the total of the source terms for X goes up to 0.01 mol/m^3 and shows the right distribution profile based on mass & light. So something on the order of dt * 0.01 mol/m^3 of X should produced at that depth location in a time step.
- When no diffusion term of X is added, then the budget more or less matches up. Apart from some small error, the budget difference is stable throughout the time series.
- Adding a diffusion of X, even in the base case of just `prod0`,  causes a big budget difference to build up.
- The problem I had been meaning to explore was which terms must be implicit terms and which must be explicit. So another switch in the script is to cast the source expressions as implicit or not. This would likely be relevant in the case where the production of X depends on its own or Y's concentration.

Running the attached script will do the following:
- it generates a base model
- iterates over the different type of source terms
- iterates over the condition of diffusion = False and True
- iterates over whether the source terms as implict = False and True (in some types of source terms, this is of course wrong).
- An interactive depth plot of the model domain and source terms is shown throughout the simulation
- At the end a time line of the relative error in the mass balance budget is plotted. This data and the plot are saved into the folder.

So my main questions are:
- How does one decide which source terms must be implicit and which explicit? There are no examples or doctests that help to show what difference this may make.
- Am I making some mistaken assumptions in my expectation of checking an order-of-magnitude mass balance?
- Why does adding a DiffusionTerm alone result in divergence of mass balance budget?

CONTENTS OF mass_balance.py
```python
from __future__ import division, print_function

import logging

logger = logging.getLogger('mass_balance')
logger.setLevel(20)
_sh = logging.StreamHandler()
_fmt = logging.Formatter('%(funcName)-20s:%(lineno)-3d : \t %(message)s', '%H:%M:%S')
_sh.setFormatter(_fmt)
logger.addHandler(_sh)

import fipy

logger.info('Using fipy version: {}'.format(fipy.__version__))

from fipy import *
from fipy.tools import numerix as np
import operator
import matplotlib.pyplot as plt

plt.ioff()
np.seterr(all='raise')


class Model(object):
    def __init__(self):

        self.vars = {}
        self.sources = {}
        self.diffusion = {}
        self.equations = {}
        self.source_totals = {}
        self.var_expected = {}
        self.budget_ = []
        self.change_sources = {}
        self.change_transport = {}
        self.artist_data = {}

    def make_domain(self, dx, nx):
        dx = PhysicalField(dx, 'mum')
        logger.info('Mesh cell size: {}'.format(dx))
        self.mesh = Grid1D(dx=dx.numericValue, nx=nx)
        logger.info('Made mesh: {}'.format(self.mesh))
        self.depths = PhysicalField(self.mesh.x() * 1000.0, 'mm')
        # mesh.x() in meters without units
        self.distances = PhysicalField(self.mesh.scaledCellDistances[:-1], unit='m')

    def add_var(self, name, unit, value, top = None, bottom = None):
        logger.info('Adding variable: {} unit: {} value: {} top: {} bottom: {}'.format(
            name, unit, value, top, bottom
            ))

        var = CellVariable(mesh=self.mesh, name=name, unit=unit, value=value, hasOld=1)
        if top is not None:
            var.constrain(top, self.mesh.facesLeft)
        if bottom is not None:
            var.constrain(bottom, self.mesh.facesRight)

        self.vars[name] = var
        self.sources[name] = []

    def add_mass(self, loc, scale, coeff):
        """
        Add an active mass distributin
        """
        C = 1.0 / np.sqrt(2 * np.pi)

        logger.info(
            'Seeding with normal profile loc: {} scale: {} coeff: {}'.format(loc, scale, coeff))

        from scipy.stats import norm
        normrv = norm(loc=loc.numericValue, scale=C ** 2 * scale.numericValue)
        val = coeff * normrv.pdf(self.depths.numericValue) * C * scale.numericValue

        mass = CellVariable(mesh=self.mesh, name='mass', unit=val.unit)
        mass.value = val
        logger.info('added mass with quantity: {}'.format(coeff))
        self.vars['mass'] = mass
        self.mass = mass
        # logger.info('mass: {}'.format(mass))

    def add_light_attenuation(self,
                              k0 = PhysicalField(15.3, '1/cm'),
                              kmass = PhysicalField(7687.5, 'cm**2/g')
                              ):

        try:
            kvar = k0 + self.mass * kmass
            kvar.inUnitsOf('1/m')
        except TypeError:
            logger.error('kvar units are not as 1/m')
            raise

        attenuation = np.cumprod(np.exp(-1.0 * kvar * self.distances))
        self.atten = CellVariable(mesh=self.mesh, name='atten')
        self.atten.value = attenuation
        self.vars['atten'] = self.atten
        # logger.info('Attenuation: {}'.format(self.atten))

    def add_light_source(self):
        """
        Unitless light intensity as % of the surface value (100)
        """
        self.light_surface = Variable(name='light0', value=0.0)
        self.light = self.light_surface * self.atten

    def add_source(self, varname, coeff, implicit = False):
        # if coeff is None:
        #     logger.info('Ignoring source of None for {}'.format(varname))
        # else:
        logger.info('Adding source for var {!r} with coeff: {!r} implicit={}'.format(
            varname, coeff, implicit
            ))
        self.sources[varname].append((coeff, implicit))

    # def add_source_from(self, varname, source_type):
    #
    #     var = sour

    def add_diffusion(self, varname, coeff):
        self.diffusion[varname] = coeff
        logger.info('Added diffusion for {}: {}'.format(varname, coeff))

    def make_var_eqn(self, name):
        logger.debug('Making eqn for var: {}'.format(name))
        var = self.vars[name]

        source_terms = []
        sources_totals = None
        for source_expr, implicit in self.sources[name]:

            if source_expr is None:
                continue

            if implicit:
                sterm = ImplicitSourceTerm(coeff=source_expr, var=var)
                # is this correct? the source expr goes as coefficient?
                # or should it somehow be refactored for var?
            else:
                sterm = source_expr

            source_terms.append(sterm)

            if sources_totals is None:
                sources_totals = source_expr
            else:
                sources_totals += source_expr

            logger.info('Added source {!r}'.format(source_expr))
            self.source_totals[name] = sources_totals

        logger.info('Full sources expr: {!r}'.format(sources_totals))

        if name in self.diffusion:

            # term_diff = ExplicitDiffusionTerm(coeff=self.diffusion[name], var=var)
            # This cannot be tested, because of the error:

            # lib/python2.7/site-packages/fipy/terms/explicitDiffusionTerm.py", line 65,
            # in _buildMatrix
            # return (var, SparseMatrix(mesh=var.mesh), b - L * var.value)

            # lib/python2.7/site-packages/fipy/matrices/scipyMatrix.py", line 321, in __mul__
            # return _ScipyMatrixFromShape.__mul__(self, other)

            # lib/python2.7/site-packages/fipy/matrices/scipyMatrix.py", line 175, in __mul__
            #     raise TypeError

            term_diff = DiffusionTerm(coeff=self.diffusion[name], var=var)
        else:
            term_diff = None

        RHS_elements = filter(None, [term_diff, ]) + source_terms

        if not RHS_elements:
            raise RuntimeError('No right hand side terms found for {} eqn'.format(name))

        RHS_terms = reduce(operator.add, RHS_elements)
        eqn = TransientTerm(var=var, coeff=1.0) == RHS_terms
        logger.info('Created {} equation: {}'.format(name, eqn))

        self.equations[name] = eqn

        return eqn

    def make_full_eqn(self):
        full_eqn = reduce(operator.and_, self.equations.values())
        logger.info('Full eqn: {}'.format(full_eqn))
        return full_eqn

    def calc_max_dt(self, ):
        D = max(self.diffusion.values())
        dx = self.depths[5] - self.depths[4]
        dt = (dx ** 2 / 2.0 / D).inUnitsOf('s')
        logger.info('Ideal dt: {}'.format(dt))
        return dt

    def integrate_over_domain(self, expr):
        # logger.warning('Integrating: {!r}'.format(expr))
        d_ = self.depths.inBaseUnits()
        expr_ = expr.inBaseUnits()
        integrated = np.trapz(expr_.numericValue, x=d_.numericValue)
        integrated = PhysicalField(integrated, expr_.unit * d_.unit)
        return integrated

    def update_mass_budget(self, dt):
        """
        Calculate the amount of mass produced/consumed from the sources and the
        amount lost due to diffusive mass transport.

        The amounts are calculated across the domain. That is the profile of the source expr for
        each variable is integrated over the depths (going from mol/m**3/s --> mol/m**2/s) and
        then multiplied with the time step to get units of mol/m**2. This is the integrated (
        areal) density of the amount produced/consumed.

        The transport loss is calculated as the amount of material leaving the domain at the
        boundaries. So this uses a quasi steady-state application of fick's law,
        and is calculated as -D * (C_edge-C-inner)/(depth_edge-depth_inner) where the inner
        refers to the cell next to the edge cell on either end of the domain. Units will be (
        m**2/s * mol/m**3 / m ==> mol/m**2/s). Multiplied by time step, it gives mol/m**2 as
        diffusive transport loss from the domain. The C_edge values of course are constrained as
        boundary conditions, and so it is only C_inner that really changes over time.

        So the budget update for the timestep is (sources_change - transport_change), which is
        the amount we expect the equation variables to change by during the time step.
        """
        logger.debug('updating mass budget for dt: {}'.format(dt))

        d_ = self.depths.inBaseUnits()

        for name, sexpr in self.source_totals.items():
            # trapz cannot handle terms with PhysicalUnit, so we recreate it with units
            change_sources = self.integrate_over_domain(sexpr)

            var = self.vars[name]
            D = self.diffusion.get(name, PhysicalField(0.0, 'm**2/s'))
            # if D is an array, then use D[0] and D[-1]

            top = -D * (var[0] - var[1]) / (d_[0] - d_[1])
            bottom = -D * (var[-1] - var[-2]) / (d_[-1] - d_[-2])
            change_transport = (top + bottom)

            # time step change
            change_net = dt * (change_sources - change_transport)
            logger.debug('Net {} change expected in timestep: {}'.format(name, change_net))

            self.var_expected[name] += change_net()

    def calculate_budget_imbalance(self):
        """
        Calculate the difference between the expected and actual amount of variable integrated in
        the domain
        """
        budget = {}
        for name in self.equations:
            expected = self.var_expected[name]
            actual = self.integrate_over_domain(self.vars[name])

            # diff = expected - actual
            budget[name] = (expected.inBaseUnits(), actual.inBaseUnits())

        return budget

    def plot_init(self, **kwargs):

        # a matplotlib figure with three domain axes: mass+light, vars, source exprs, source totals
        # a time series plot of the mass balance budget

        self.fig1, axes = plt.subplots(nrows=1, ncols=4, sharey=True, **kwargs)
        self.axes = list(axes)
        self.axMass, self.axVars, self.axTotals, self.axSources = axes

        # plotting depth in mm
        DEPTHS = self.depths.numericValue * 1000
        zeros = np.zeros_like(DEPTHS)

        line = self.axVars.plot(zeros, DEPTHS, label='X')[0]
        self.artist_data[line] = self.vars['X']
        line.normalize_data_ = True

        # self.axVarsb = plt.twiny(self.axVars)
        # line = self.axVarsb.plot(zeros, DEPTHS, color='C2', label='Y')[0]
        # self.artist_data[line] = self.vars['Y']
        # line.normalize_data_ = True
        # self.axes.append(self.axVarsb)

        for name in self.equations:

            total = self.source_totals.get(name)
            if total is not None:
                line = self.axTotals.plot(zeros, DEPTHS, label='{} source_total'.format(name))[0]
                self.artist_data[line] = total

        for name, sources in self.sources.items():
            for n, (source_expr, implicit) in enumerate(sources):
                if source_expr is None:
                    continue
                label = '{}:{}'.format(name, n)
                line = self.axSources.plot(zeros, DEPTHS, label=label)[0]
                self.artist_data[line] = source_expr
                logger.info('Added line {} for {!r}'.format(label, source_expr))

        self.axVars.invert_yaxis()
        self.axVars.set_title('Vars')
        self.axTotals.set_title('Totals')
        self.axSources.set_title('Sources')
        self.axMass.set_title('Mass')

        line = self.axMass.plot(self.mass.numericValue, DEPTHS, label='mass')[0]
        # a plot for the light
        self.axLight = self.axMass.twiny()
        self.axLight.set_xscale('log')
        self.axLight.set_xlim(0.1, 100)
        line = self.axLight.plot(zeros, DEPTHS, 'r--', lw=1.25, label='light')[0]
        self.artist_data[line] = self.light

        self.clockstr = '{0:02d}h {1:02d}m {2:03d}s'
        self.clock_artist = plt.annotate(self.clockstr.format(0, 0, 0),
                                         xy=(0.01, 0.01),
                                         xycoords='figure fraction',
                                         size='medium',
                                         color='r')

        for ax in axes:
            ax.legend(loc=0)
            ax.autoscale(True, axis='x')

    def plot_update(self, TIME):

        H, M, S = [int(s.value) for s in TIME.inUnitsOf('h', 'min', 's')]
        self.clock_artist.set_text(self.clockstr.format(H, M, S))

        for line, data in self.artist_data.items():

            ax = line.axes

            normit = getattr(line, 'normalize_data_', False)
            if normit:
                logger.warning('Normalizing data for {}'.format(line.get_label()))
                data, span = normalize(data)
                ax.set_xlabel('{:.2} {}'.format(span.value, span.unit.name()))

            data = data.inBaseUnits()
            if hasattr(data, 'unit'):
                unit = data.unit.name()
                ax.set_xlabel(unit)
                data = data.numericValue

            line.set_xdata(data)

            # if ax is self.axVars:
            #     logger.info('vars data: {} {}'.format(min(data), max(data)))

        for ax in self.axes:
            ax.relim()
            ax.autoscale_view()

        self.fig1.canvas.draw_idle()
        self.fig1.canvas.flush_events()


def inhibition(x, Kmax, Khalf):
    return (Kmax - x) / (2 * Kmax - Khalf - x) * (x < Kmax)


def saturation(x, Km):
    return x / (Km + x)


def optimum(x, Ks, Ki):
    return x / (x + Ks) / (1 + x / Ki)


def consumption(model, Q):
    """ static consumption in domain"""
    return Q


def production(model, Q):
    "photochemical production by mass"
    return model.mass * Q * optimum(model.light, 1.0, 10.0)


def normalize(x):
    span = (max(x) - min(x))
    if np.allclose(span.numericValue, 0.0):
        return x, span
    else:
        return (x - min(x)) / span, span


def with_responses(model, base, response_defs = ()):
    responses_ = []
    if response_defs:
        for response_def in response_defs:
            logger.info('Creating response from: {!r}'.format(response_def))
            varname, resp_type, params = response_def
            # logger.info('Defining {} response to {}: {}'.format(resp_type, varname, params))
            var = model.vars[varname]
            resp_func = globals()[resp_type]
            resp = resp_func(var, *params)
            responses_.append(resp)
            logger.info('Created {} response: {!r}'.format(resp_type, resp))

        return base * reduce(operator.mul, responses_)

    else:
        return base


def make_model(diffusion = True):
    logger.info('Making base model')
    model = Model()
    DX = 100
    model.make_domain(dx=100, nx=100)

    # X var
    model.add_var('X', 'mol/m**3', 1e-30, 1e-30, 1e-30)
    # y var
    model.add_var('Y', 'mol/m**3', 0.0, 0.0, 1e-3)

    # active mass
    model.add_mass(loc=PhysicalField(2, 'mm'), scale=PhysicalField(1.5, 'mm'),
                   coeff=PhysicalField(12, 'mg/cm**3'))
    model.add_light_attenuation()
    model.add_light_source()

    # add diffusion
    Dx = PhysicalField(0.03, 'cm**2/h')
    Dy = PhysicalField(0.02, 'cm**2/h')
    model.add_diffusion('Y', Dy)

    if diffusion:
        model.add_diffusion('X', Dx)

    return model


#################################################################
##############  MODEL DEFINITIONS
#################################################################

Q1 = PhysicalField(0.00841, 'mol/g/h')
X_Kmax = PhysicalField(10e-6, 'mol/m**3')
X_Khalf = PhysicalField(0.7e-3, 'mol/m**3')

Y_Kmax = PhysicalField(0.5e-3, 'mol/m**3')
Y_Khalf = PhysicalField(0.3e-3, 'mol/m**3')

Q2 = PhysicalField(-2e-3, 'mol/l/h')
Q2_X_Km = PhysicalField(1e-5, 'mol/l')


# Q2 = PhysicalField(-0.0012, 'mol/g/h')

# add sources
def get_source_defs(source_type, model):
    logger.info('Defining source_types for model')
    MODEL_TYPES = dict(
        [
            ('production', {'X': (production(model, Q1),),
                            'Y': (None,)}  # no sources, pure diffusion equation
             ),

            ('production_consume', dict(X=(
                production(model, Q1),
                consumption(model, Q2),
                ),
                Y=(None,),  # just diffusion
                ),
             ),

            ('production:inhibit[X]', dict(X=(
                with_responses(model,
                               base=production(model, Q1),
                               response_defs=[
                                   ('X', 'inhibition', (X_Kmax, X_Khalf)),
                                   ]),
                ),
                Y=(None,)  # just diffusion
                )
             ),

            ('production:inhibit[X,Y]', dict(X=(
                with_responses(model,
                               base=production(model, Q1),
                               response_defs=[
                                   ('X', 'inhibition', (X_Kmax, X_Khalf)),
                                   ('Y', 'inhibition', (Y_Kmax, Y_Khalf)),
                                   ]),
                ),
                Y=(None,),  # no sources, pure diffusion equation
                )
             ),

            ('production:inhibit[X,Y]_consume', dict(X=(
                with_responses(model,
                               base=production(model, Q1),
                               response_defs=[
                                   ('X', 'inhibition', (X_Kmax, X_Khalf)),
                                   ('Y', 'inhibition', (Y_Kmax, Y_Khalf)),
                                   ]),

                consumption(model, Q2),
                ),
                Y=(None,),  # no sources, pure diffusion equation
                ),
             ),

            ('production:inhibit[X,Y]_consume:saturate[X]', dict(X=(
                with_responses(model,
                               base=production(model, Q1),
                               response_defs=[
                                   ('X', 'inhibition', (X_Kmax, X_Khalf)),
                                   ('Y', 'inhibition', (Y_Kmax, Y_Khalf)),
                                   ]),

                with_responses(model,
                               base=consumption(model, Q2),
                               response_defs=[
                                   ('X', 'saturation', (Q2_X_Km,)),
                                   ],
                               ),
                ),
                Y=(None,),  # no sources, pure diffusion equation
                )
             )

            ])
    return MODEL_TYPES[source_type]


RES_LIM = 1e-8
MAX_SWEEPS = 20


def run_full_model(source_type, diffusion = True, implicit = False, confirm = True):
    model = make_model(diffusion=diffusion)
    source_defs = get_source_defs(source_type, model)
    logger.info('Running full model for source_type: {}'.format(source_type))

    for varname, sources in source_defs.items():
        for source in sources:
            logger.info('Adding source for {}: {!r}'.format(varname, source))
            model.add_source(varname, source, implicit=implicit)

    for varname in source_defs:
        model.make_var_eqn(varname)

    meqn = model.make_full_eqn()

    from fipy.solvers.scipy import DefaultSolver
    solver = DefaultSolver()
    logger.info("using solver: {}".format(solver))

    # determine ideal dT
    dT = model.calc_max_dt() / 2.0  # take half the max dT
    TOTAL = PhysicalField(0.5, 'h')
    time_ = Variable(unit='h', name='time')
    LIGHT_SET = True
    model.light_surface.value = 100.0
    # logger.warning('Light: {}'.format(model.light))

    nsteps = int(TOTAL / dT)
    logger.info('Solving for {} with dT={} nsteps = {}'.format(TOTAL, dT, nsteps))

    model.plot_init(figsize=(8, 8))
    model.plot_update(time_)
    plt.suptitle('Diffusion(X) = {} Implicit = {}'.format(diffusion, implicit))
    plt.show(block=False)

    if confirm:
        rinp = raw_input('Start solution?  ')
        if rinp.strip().lower() not in ('', 'y'):
            logger.warning('No selected')
            return

    logger.info('starting solution')

    # collect the time series of budget
    TIMES = []
    BUDGET_X = []

    for step in range(nsteps):

        for varname in model.equations:
            var = model.vars[varname]
            var.updateOld()
            logger.debug('Updated old of var {}'.format(varname))

        res = 1
        nsweeps = 0

        while (res > RES_LIM) and (nsweeps < MAX_SWEEPS):
            res = meqn.sweep(solver=solver, dt=float(dT.numericValue))
            nsweeps += 1

        if step == 0:
            # set up the expected value as the current one after the first timestep
            # this is the 'initial condition' of the variable
            for name in model.equations:
                var = model.vars[name]
                var.updateOld()
                model.var_expected[name] = model.integrate_over_domain(var)

        model.update_mass_budget(dT)

        logger.debug('Step #{} res: {:.3g} sweeps={}'.format(step, res, nsweeps))
        if step % 10 == 0:
            print('\n')
            logger.info('Step #{} res: {:.3g} sweeps={} time={:.6} of {:.6}'.format(step, res,
                                                                                    nsweeps,
                                                                                    time_,
                                                                                    TOTAL))
            budget = model.calculate_budget_imbalance()

            name = 'X'
            expected, actual = budget[name]
            diff = expected - actual
            relative_imbalance = float((diff / actual) * 100.0)
            logger.info("\n\n{} imbalance: {:.2f}% \nexpected= {} \nactual= {} \ndifference= {"
                        "}".format(
                name, relative_imbalance, expected, actual, diff
                ))

            TIMES.append(float(time_.numericValue))
            BUDGET_X.append(relative_imbalance)

            model.plot_update(time_)
            # this should maybe come after updating time value below, but not critical

        time_.value = time_.value + dT

        # after half the time switch on the light
        if time_() > TOTAL / 2.0 and (LIGHT_SET):
            model.light_surface.value = 0.0
            LIGHT_SET = False
            logger.warning('Light switched to {} at {}'.format(LIGHT_SET, time_))

    plt.close()

    plt.plot(TIMES, BUDGET_X, ':.')
    plt.xlabel('Time (s)')
    plt.ylabel('Relative error (%)')
    plt.title('Diffusion(X) = {} Implicit = {}'.format(DIFFUSION, IMPLICIT))
    fname = 'source[{}]_diffusion[{}]_implicit[{}]'.format(SOURCE_TYPE, DIFFUSION, IMPLICIT,)
    plt.savefig(fname + '.png')
    plt.close()

    np.savetxt(fname + '.txt', (TIMES, BUDGET_X))

    # plt.show()


if __name__ == '__main__':

    SOURCE_TYPES = (
        'production',
        'production_consume',
        'production:inhibit[X]',
        'production:inhibit[X,Y]',
        'production:inhibit[X,Y]_consume',
        'production:inhibit[X,Y]_consume:saturate[X]',
        )

    IMPLICIT = False
    DIFFUSION = True
    SOURCE_TYPE = 'production'

    try:
        for SOURCE_TYPE in SOURCE_TYPES:
            for IMPLICIT in (False, True):
                for DIFFUSION in (False, True):
                    run_full_model(source_type=SOURCE_TYPE,
                                   diffusion=DIFFUSION,
                                   implicit=IMPLICIT,
                                   confirm=False)
    except KeyboardInterrupt:
        logger.warning('User interrupt')

```

**Comment by guyer** [comment]
This needs to be simplified dramatically before I can help. 

Please provide an example of a full FiPy equation that you believe doesn't conserve and the equivalent math of what you're trying to solve.

**Comment by achennu** [comment]
Agreed. I have been using the supplied code to dig into the problem. I think I have resolved the issue now. I will come back with more info shortly.

**Comment by achennu** [comment]
@guyer Sorry for the lapse. I have explored this and think resolved it.

So the issue was linked to #553 

The budget balance of the variables works out, as long as the spatial dimensions (in units) matches the unit of the mesh grid. That is, if the mesh is created as meters (without units), then all the equation term coefficients need to be converted to meters, since the mesh does not itself take units into account on creation. A workaround I used for this was to create the mesh in meters, and then convert all parameters for the equation terms `inBaseUnits`, which converts it into meters.

I will update #553 with a couple of notes. Thanks for closing this issue.
